src/murmur/MurmurGRPCImpl.cpp

```diff
@@ -138,10 +140,18 @@ MurmurRPCImpl::MurmurRPCImpl(const QString &address, std::shared_ptr<::grpc::Ser
 	m_completionQueue = builder.AddCompletionQueue();
 	m_server = builder.BuildAndStart();
 	meta->connectListener(this);
+	b_IsRunning = true;
 	start();
 }
 
 MurmurRPCImpl::~MurmurRPCImpl() {
+	void *ignored_tag;
+	bool ignored_ok;
+	b_IsRunning = false;
```

Comment:
Even with this ::run might still pick up another tag and start processing it or it might be processing one right now. Then the destructor might return before processing in ::run is complete.

TBO looking at the shutdown sequence I'm wondering whether it is still safe to process anything at all at the point of this destructor or whether grpc shutdown would have to happen before server shutdowns to be safe. Did you happen to investigate that?

Reply:
I did investigate a little about this. Yes, there is a problem here, that I decided not to fix as the server is shutting down anyway.

Say you have something listening for events using the streaming protocols. If you you set the timeout to something like 10s in the future, call stop on the server, and don't stop the main loop (hoping to drain it there) it invokes the oom killer. If you don't set a deadline, and try to wait for the server to finish with m_server->Wait(), it hangs. Nothing is really able to be completed at this point in the process. You can also call m_server->Wait() after requesting the shutdown, it just causes it to hang until the deadline hits and everything is disconnected.

TL;DR: Yes, it will try to do one more action, but that action will fail because the server shutdown has been requested and the deadline is in the past.